### DISCUSSION TITLE
`Lint/ShadowingOuterLocalVariable` false positive

### DISCUSSION THREAD

**Discussion by DannyBen** [rubocop/rubocop#11495] [Q&A]
I guess this is nitpicking, but the `Lint/ShadowingOuterLocalVariable` complains about something that I believe is valid.

```ruby
def run(argv = [])
  if some_condition
    userfile = Userfile.new
    userfile.context = :local
    return userfile.run
  end

  external_files.each do |userfile|
    return userfile.run(argv) if some_condition
  end
end
```

I know this can be written more nicely, to avoid this, but I wonder if `userfile` should be considered "outer variable" if it is a) in a block, and b) that block always returns.

As I see it, I am not shadowing anything - the `userfile` variable is used independently in two different blocks, and will never intersect.

**Comment by airled** [comment]
As [said](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ShadowingOuterLocalVariable) shadowing is about using name of a variable.
`If` does not create its own scope.
```ruby
irb(main):001:1* if 5 == 5
irb(main):002:1*   a = 10
irb(main):003:0> end
=> 10
irb(main):004:0> a
=> 10
```
So your variable `userfile = Userfile.new` becomes function-visible. And then you use block with inner `userfile` variable inside a block that already knows about outer `userfile`. That's why rubocop gives you the warning.
